File: android/biangbiang-ui/src/main/java/dev/veeso/biangbiangui/ui/BiangBiangContext.kt
package dev.veeso.biangbiangui.ui

import androidx.compose.runtime.Composable
import androidx.compose.runtime.State
import androidx.compose.runtime.derivedStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.staticCompositionLocalOf
import androidx.lifecycle.compose.collectAsStateWithLifecycle
import dev.veeso.biangbiangui.config.BiangBiangConfig
import dev.veeso.biangbiangui.config.LanguageProfile
import dev.veeso.biangbiangui.config.LanguageVariant
import dev.veeso.biangbiangui.protocols.AudioProvider
import dev.veeso.biangbiangui.protocols.Transliterator
import dev.veeso.biangbiangui.services.SettingsStore
import dev.veeso.biangbiangui.services.TextProcessingEngine

/**
 * App-wide context exposed to every library screen.
 *
 * Mirrors iOS `BiangBiangContext` (`@Observable` env object: `config`,
 * `settings`, `engine` rebuildable via `rebuildEngine()`, `audio`,
 * `activeProfile`, `activeVariant`, `variant(forId:)`).
 *
 * ## Android divergence vs. iOS
 *
 * iOS holds `engine` as an `@Observable` mutable property and calls
 * `rebuildEngine()` imperatively after `selectedVariantId` changes. On Android
 * `SettingsStore` is Flow-based, so the equivalent surface is reactive: the
 * holder collects `settings.selectedVariantId` and derives `activeVariant` and
 * `engine` from it. Reading `engine`/`activeVariant` inside Compose
 * recomposes when the selected variant changes — there is no imperative
 * `rebuildEngine()` to call, the rebuild is driven by the Flow (the documented
 * iOS `rebuildEngine()` contract, expressed idiomatically).
 *
 * Ownership: [BiangBiangRoot] builds one holder and provides it via
 * [LocalBiangBiangContext]. Screens only read it.
 */
class BiangBiangContext(
    val config: BiangBiangConfig,
    val settings: SettingsStore,
    /** Resolved audio provider (plugin override or system TTS). */
    val audio: AudioProvider,
    /** Reactive selected-variant id, collected from [SettingsStore]. */
    private val selectedVariantId: State<String>,
) {
    /** Pass-through transliterator used when no active variant supplies one. */
    private val identityTransliterator = Transliterator { it }

    /**
     * The first language profile in the config. Apps that support a single
     * script always have exactly one profile; this is its canonical accessor.
     * Mirrors iOS `activeProfile`.
     */
    val activeProfile: LanguageProfile
        get() {
            check(config.languages.isNotEmpty()) {
                "BiangBiangConfig must declare at least one LanguageProfile"
            }
            return config.languages[0]
        }

    /**
     * The variant whose `id` matches `settings.selectedVariantId`, falling back
     * to the first variant of the active profile. Mirrors iOS `activeVariant`.
     * Backed by [derivedStateOf] so reads recompose on variant change.
     */
    val activeVariant: LanguageVariant?
        get() = activeVariantState.value

    private val activeVariantState: State<LanguageVariant?> = derivedStateOf {
        val id = selectedVariantId.value
        activeProfile.variants.firstOrNull { it.id == id }
            ?: activeProfile.variants.firstOrNull()
    }

    /**
     * The active [TextProcessingEngine], derived from the active profile's
     * `scriptRanges` and the active variant's `transliterator`. Recomputed
     * whenever the selected variant changes — this is the Android equivalent of
     * iOS `rebuildEngine()`.
     */
    val engine: TextProcessingEngine
        get() = engineState.value

    private val engineState: State<TextProcessingEngine> = derivedStateOf {
        TextProcessingEngine(
            scriptRanges = activeProfile.scriptRanges,
            transliterator = activeVariantState.value?.transliterator
                ?: identityTransliterator,
        )
    }

    /**
     * Searches every profile's variant list for [id]. Used by History and TTS
     * screens that need to resolve a stored `variantId` back to a variant.
     * Mirrors iOS `variant(forId:)`.
     */
    fun variant(forId: String): LanguageVariant? =
        config.languages.asSequence()
            .flatMap { it.variants.asSequence() }
            .firstOrNull { it.id == forId }
}

/**
 * The injected [BiangBiangContext]. [BiangBiangRoot] provides it; every screen
 * resolves it with `LocalBiangBiangContext.current`. Mirrors iOS
 * `@Environment(BiangBiangContext.self)`.
 */
val LocalBiangBiangContext = staticCompositionLocalOf<BiangBiangContext> {
    error("BiangBiangContext not provided. Wrap content in BiangBiangRoot.")
}

/**
 * Builds and remembers a [BiangBiangContext], collecting the reactive
 * `selectedVariantId` from [settings]. Called once from [BiangBiangRoot].
 */
@Composable
internal fun rememberBiangBiangContext(
    config: BiangBiangConfig,
    settings: SettingsStore,
    audio: AudioProvider,
): BiangBiangContext {
    val variantId = settings.selectedVariantId
        .collectAsStateWithLifecycle(
            initialValue = config.languages.firstOrNull()
                ?.variants?.firstOrNull()?.id ?: "",
        )
    return remember(config, settings, audio) {
        BiangBiangContext(config, settings, audio, variantId)
    }
}

File: Sources/BiangBiangUI/UI/BiangBiangContext.swift
//
//  BiangBiangContext.swift
//  BiangBiangUI
//
//  App-wide context injected into the SwiftUI environment. Mirrors the
//  pattern in the reference app's `@EnvironmentObject AppSettings` +
//  `@Environment(AudioPlayerService.self)`, unified into a single object.
//

import Observation
import SwiftUI

// MARK: - Identity transliterator (fallback when no variant is active)

/// Pass-through transliterator used when no active variant supplies one.
/// Returns its input unchanged so the engine still produces output.
private struct IdentityTransliterator: Transliterator {
    func transliterate(_ scriptSpan: String) -> String {
        scriptSpan
    }
}

// MARK: - BiangBiangContext

/// Shared context that every library screen reads from `@Environment`.
///
/// Ownership: the consuming app creates one `BiangBiangContext` in its root
/// view and injects it with `.environment(context)`.  Screens never own or
/// create it; they only read it.
@MainActor
@Observable
public final class BiangBiangContext {
    // MARK: - Public stored properties

    public let config: BiangBiangConfig
    public let settings: SettingsStore
    /// The resolved audio provider (plugin override or system TTS).
    /// Resolution happens in the root view that constructs this context.
    public let audio: any AudioProvider

    // MARK: - Observable engine

    /// Rebuilt whenever `selectedVariantId` changes via `rebuildEngine()`.
    public private(set) var engine: TextProcessingEngine

    // MARK: - Init

    /// Creates the context and builds an initial `TextProcessingEngine` from
    /// the active profile / variant at construction time.
    public init(config: BiangBiangConfig, settings: SettingsStore, audio: any AudioProvider) {
        self.config = config
        self.settings = settings
        self.audio = audio
        // Build engine inline; `activeProfile` / `activeVariant` are pure
        // computed properties so they are safe to call before `self` is fully
        // initialised (no stored-property cycles).
        let profile = config.languages.first
        let variantId = settings.selectedVariantId
        let variant = profile?.variants.first(where: { $0.id == variantId })
            ?? profile?.variants.first
        let transliterator: any Transliterator = variant?.transliterator ?? IdentityTransliterator()
        engine = TextProcessingEngine(
            scriptRanges: profile?.scriptRanges ?? [],
            transliterator: transliterator
        )
    }

    // MARK: - Derived helpers

    /// The first language profile in the config. Apps that support a single
    /// script always have exactly one profile; this is its canonical accessor.
    public var activeProfile: LanguageProfile {
        precondition(!config.languages.isEmpty,
                     "BiangBiangConfig must declare at least one LanguageProfile")
        return config.languages[0]
    }

    /// The variant whose `id` matches `settings.selectedVariantId`, falling
    /// back to the first variant of the active profile when there is no match
    /// (e.g. first launch before any selection has been persisted).
    public var activeVariant: LanguageVariant? {
        let id = settings.selectedVariantId
        return activeProfile.variants.first(where: { $0.id == id })
            ?? activeProfile.variants.first
    }

    /// Rebuilds `engine` from the active profile's `scriptRanges` and the
    /// active variant's `transliterator`. Call this after `selectedVariantId`
    /// changes (e.g. inside an `onChange(of:)` modifier on the root view).
    public func rebuildEngine() {
        let transliterator: any Transliterator =
            activeVariant?.transliterator ?? IdentityTransliterator()
        engine = TextProcessingEngine(
            scriptRanges: activeProfile.scriptRanges,
            transliterator: transliterator
        )
    }

    /// Searches every profile's variant list for `id`. Used by History and
    /// TTS screens that need to resolve a stored `variantId` string back to
    /// a `LanguageVariant`.
    public func variant(forId id: String) -> LanguageVariant? {
        config.languages.lazy
            .flatMap(\.variants)
            .first(where: { $0.id == id })
    }
}
